pkg/hub/client/local/client_test.go

```diff
@@ -0,0 +1,274 @@
+package local
+
+import (
+	"context"
+	"os"
+	"regexp"
+	"strings"
+	"testing"
+
+	"capact.io/capact/internal/cli/heredoc"
+	"capact.io/capact/internal/ptr"
+	gqllocalapi "capact.io/capact/pkg/hub/api/graphql/local"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TODO(review): THIS FILE WILL BE REMOVED BEFORE MERGING. IT WAS ADDED ONLY FOR DEMO/PR TESTING PURPOSES.
+
+const (
+	RunValidation       = "RUN_VALIDATION"
+	typeInstanceTypeRef = "cap.type.testing:0.1.0"
+)
+
+type StorageSpec struct {
+	URL           *string `json:"url,omitempty"`
+	AcceptValue   *bool   `json:"acceptValue,omitempty"`
+	ContextSchema *string `json:"contextSchema,omitempty"`
+}
+
+// Prerequisite:
+//   Before running this test, make sure that Local Hub is running:
+//     cd hub-js; APP_NEO4J_ENDPOINT=bolt://localhost:7687 APP_NEO4J_PASSWORD=okon APP_HUB_MODE=local npm run dev
+//
+// To run this test, execute:
+// RUN_VALIDATION=true go test ./pkg/hub/client/local/ -v -count 1
+func TestExternalStorageInputValidation(t *testing.T) {
+	run := os.Getenv(RunValidation)
+	if run == "" {
+		t.Skipf("skipping running example test as the env %s is not provided", RunValidation)
+	}
+
+	ctx := context.Background()
+	cli := NewDefaultClient("http://localhost:8080/graphql")
+
+	tests := map[string]struct {
+		// given
+		storageSpec interface{}
+		value       map[string]interface{}
+		context     interface{}
+
+		// then
+		expErrMsg string
+	}{
+		"Should rejected value": {
+			storageSpec: StorageSpec{
+				URL:         ptr.String("http://localhost:5000/fake"),
+				AcceptValue: ptr.Bool(false),
+			},
+			value: map[string]interface{}{
+				"name": "Luke Skywalker",
+			},
+			expErrMsg: heredoc.Doc(`
+              while executing mutation to create TypeInstance: All attempts fail:
+              #1: graphql: failed to create TypeInstance: failed to create the TypeInstances: 2 error occurred:
+              	* Error: External backend "MOCKED_ID": input value not allowed
+              	* Error: rollback externally stored values: External backend "MOCKED_ID": input value not allowed`),
```

Comment:
As I understand, one of the errors is from `Store` and the second is from `rollbackExternalStoreAction`.  Is it an option to not perform a rollback in case of a validation error?

Reply:
I tried to but because we can create in a single mutation more than one TypeInstance it complicates the solution, so I decide to give up.